src/Server/BaseServer.ts

```diff
@@ -313,9 +314,55 @@ export abstract class BaseServer implements IServer {
   static fromJSONBase<T extends BaseServer>(value: IReviverValue, ctor: new () => T, keys: readonly (keyof T)[]): T {
-    const result = Generic_fromJSON(ctor, value.data, keys);
-    result.savedScripts = value.data.runningScripts;
-    return result;
+    const server = Generic_fromJSON(ctor, value.data, keys);
+    server.savedScripts = value.data.runningScripts;
+    // Possibly migrate to using maps for scripts and textfiles.
+    // This is done here, directly at load, instead of the usual upgrade
+    // logic, for two reasons:
+    // 1) Our utility functions depend on it, so the upgrade logic itself needs
+    //    the data to be in maps, even the logic written earlier than 2.3!
+    // 2) If the upgrade logic throws, and then you soft-reset at the recovery screen (or
+    //    maybe don't even see the recovery screen), you can end up with a "migrated"
+    //    save that still has arrays.
+    const newDirectory = resolveDirectory("v2.3FileChanges/") as Directory;
+    let invalidScriptCount = 0;
+    // There was a brief dev window where Server.scripts was already a map but the filepath changes weren't in yet.
+    // Thus, we can't skip this logic just because it's already a map.
+    const oldScripts = Array.isArray(server.scripts) ? (server.scripts as Script[]) : [...server.scripts.values()];
+    server.scripts = new JSONMap();
+    // In case somehow there are previously valid filenames that can't be sanitized, they will go in a new directory with a note.
+    for (const script of oldScripts) {
+      let newFilePath = resolveScriptFilePath(script.filename);
+      if (!newFilePath) {
+        newFilePath = `${newDirectory}script${++invalidScriptCount}.js` as ScriptFilePath;
+        script.content = `// Original path: ${script.filename}. Path was no longer valid\n` + script.content;
+      }
+      script.filename = newFilePath;
+      server.scripts.set(newFilePath, script);
+    }
+    let invalidTextCount = 0;
+    // Handle changing textFiles to a map as well as FilePath changes at the same time.
+    if (Array.isArray(server.textFiles)) {
+      const oldTextFiles = server.textFiles as (TextFile & { fn?: string })[];
+      server.textFiles = new JSONMap();
+      for (const textFile of oldTextFiles) {
+        const oldName = textFile.fn ?? textFile.filename;
+        delete textFile.fn;
+
+        let newFilePath = resolveTextFilePath(oldName);
+        if (!newFilePath) {
+          newFilePath = `${newDirectory}text${++invalidTextCount}.txt` as TextFilePath;
+          textFile.content = `// Original path: ${textFile.filename}. Path was no longer valid\n` + textFile.content;
+        }
+        textFile.filename = newFilePath;
+        server.textFiles.set(newFilePath, textFile);
+      }
+    }
+    if (invalidScriptCount || invalidTextCount) {
+      // If we had to migrate names, don't run scripts for this server.
+      server.savedScripts = [];
+    }
+    return server;
   }
```

Comment:
Isn't all this migration code going to be ran every single time the game is loaded, instead of just when the version warrants the migration code to run?

We can choose to only run the migration code if textFiles is still an array, because if textFiles is a map already then the migration has already occurred.

Reply:
Yes, that was intentional. Because of various things like the different timing of introduction of maps vs path-checking code, I wasn't sure that there was a safe "it's definitely been correctly migrated" signal.

However, the presence of a map for text files is *probably* good enough.

One thing this whole migration has shown me is that we're *definitely* better off using the approach I took for RunningScripts: leave the save format the same (a simple array), and transform it to a map on save/load. This keeps backwards/forwards compatibility and just generally saves a lot of headaches.